### PR TITLE
Faster China

### PR DESCRIPTION
**Pull request by nirLumigo**
- Cache kinesis
- Cancel http cache in China
- Notify user if spans aren't sent to Kinesis
- Change configuration to `max_attempts=1`

### CODE REVIEW

File: src/lumigo_tracer/lumigo_utils.py [hunk 1]
```diff
@@ -20,6 +20,11 @@
 except Exception:
     boto3 = None
 
+try:
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:23]: merge with the other try/except

File: src/lumigo_tracer/lumigo_utils.py [hunk 2]
```diff
@@ -67,10 +72,20 @@
 
 _logger: Dict[str, logging.Logger] = {}
 
+EDGE_KINESIS_BOTO_CLIENT = None
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:75]: not const

File: src/lumigo_tracer/lumigo_utils.py [hunk 3]
```diff
@@ -324,9 +336,32 @@ def _publish_spans_to_kinesis(to_send: bytes, region: str) -> int:
...
+        get_logger().exception("Failed to send spans to Kinesis", exc_info=err)
+        warn_client(f"Failed to send spans to Kinesis: {err}")
     return int((time.time() - start_time) * 1000)
 
 
+def _get_edge_kinesis_boto_client(region: str, aws_access_key_id: str, aws_secret_access_key: str):
+    global edge_kinesis_boto_client
+    if (
+        not edge_kinesis_boto_client
+        or os.environ.get("LUMIGO_KINEISIS_INITIALIZATION_CONNECTION", "").lower()  # noqa
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:349]: can you extract it to a util function?

Comment (saartochner) [reply src/lumigo_tracer/lumigo_utils.py:349]: why do you have this key at all?
you're not initializing during cold start, so this flag is irrelevant.
In your implementation, this flag is actually "should_reuse_connection", which is different :)

File: src/lumigo_tracer/lumigo_utils.py [hunk 4]
```diff
@@ -324,9 +336,32 @@ def _publish_spans_to_kinesis(to_send: bytes, region: str) -> int:
...
             aws_secret_access_key=Configuration.edge_kinesis_aws_secret_access_key,
         )
+    except Exception as err:
+        get_logger().exception("Failed to send spans to Kinesis", exc_info=err)
+        warn_client(f"Failed to send spans to Kinesis: {err}")
     return int((time.time() - start_time) * 1000)
 
 
+def _is_edge_kinesis_connection_cache_disabled() -> bool:
+    return os.environ.get("LUMIGO_KINEISIS_INITIALIZATION_CONNECTION", "").lower() == "false"
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:346]: check out my other comment - this is a wrong name

File: src/lumigo_tracer/lumigo_utils.py (hunk 3 above)
Comment (nirLumigo) [reply src/lumigo_tracer/lumigo_utils.py]: Ok, I'll rename the env variable

File: src/lumigo_tracer/lumigo_utils.py [hunk 5]
```diff
@@ -324,9 +336,32 @@ def _publish_spans_to_kinesis(to_send: bytes, region: str) -> int:
...
+
+
+def _get_edge_kinesis_boto_client(region: str, aws_access_key_id: str, aws_secret_access_key: str):
+    global edge_kinesis_boto_client
+    if not edge_kinesis_boto_client or _is_edge_kinesis_connection_cache_disabled():
+        get_logger().info("Creating new boto3 connection")
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:352]: unecessary IMO

File: src/test/unit/test_tracer.py [hunk 6]
```diff
@@ -419,9 +419,9 @@ def lambda_test_function(event, context):
     span_sent = json.loads(records[1]["Data"].decode())[0]
     assert span_sent["event"] == json.dumps(event)
     # Used the client from the decorator params
-    boto3.client.assert_called_with(
```
Comment (saartochner) [inline src/test/unit/test_tracer.py:422]: ?

**Review by saartochner — APPROVED** [review]

**Pull request merged**